verticapy/learn/tsa/__init__.py

```diff
@@ -44,3 +44,1801 @@
+                if len(name) != len(X)
+                else name[idx]
+            )
+            all_pred += ["{} AS {}".format(self.deploySQL()[idx], name_tmp)]
+            transform_relation = transform_relation.replace("[X{}]".format(idx), elem)
+        columns = vdf.get_columns() + all_pred
+        relation = vdf.__genSQL__()
+        for i in range(nlead):
+            query = "SELECT ({} - LAG({}, 1) OVER (ORDER BY {}))::VARCHAR FROM {} ORDER BY {} DESC LIMIT 1".format(
+                ts, ts, ts, relation, ts
+            )
+            deltat = vdf._VERTICAPY_VARIABLES_["cursor"].execute(query).fetchone()[0]
+            query = "SELECT (MAX({}) + '{}'::interval)::VARCHAR FROM {}".format(
+                ts, deltat, relation
+            )
+            next_t = vdf._VERTICAPY_VARIABLES_["cursor"].execute(query).fetchone()[0]
+            if i == 0:
+                first_t = next_t
+            new_line = "SELECT '{}'::TIMESTAMP AS {}, {}".format(
+                next_t,
+                ts,
+                ", ".join(
+                    [
+                        "NULL AS {}".format(column)
+                        for column in vdf.get_columns(exclude_columns=[ts])
+                    ]
+                ),
+            )
+            relation_tmp = "(SELECT {} FROM {} UNION ALL ({})) VERTICAPY_SUBTABLE".format(
+                ", ".join([ts] + vdf.get_columns(exclude_columns=[ts])),
+                relation,
+                new_line,
+            )
+            query = "SELECT {} FROM {} ORDER BY {} DESC LIMIT 1".format(
+                ", ".join(self.deploySQL()), transform_relation.format(relation_tmp), ts
+            )
+            prediction = vdf._VERTICAPY_VARIABLES_["cursor"].execute(query).fetchone()
+            for idx, elem in enumerate(X):
+                prediction[idx] = "{} AS {}".format(prediction[idx], elem)
+            columns_tmp = vdf.get_columns(exclude_columns=[ts] + X)
+            new_line = "SELECT '{}'::TIMESTAMP AS {}, {} {}".format(
+                next_t,
+                ts,
+                ", ".join(prediction),
+                (", " if (columns_tmp) else "")
+                + ", ".join(["NULL AS {}".format(column) for column in columns_tmp]),
+            )
+            relation = "(SELECT {} FROM {} UNION ALL ({})) VERTICAPY_SUBTABLE".format(
+                ", ".join([ts] + X + vdf.get_columns(exclude_columns=[ts] + X)),
+                relation,
+                new_line,
+            )
+        final_relation = "(SELECT {} FROM {}) VERTICAPY_SUBTABLE".format(
+            ", ".join(columns), transform_relation.format(relation)
+        )
+        result = vdf_from_relation(final_relation, "VAR", self.cursor,)
+        if nlead > 0:
+            for elem in X:
+                result[elem].apply(
+                    "CASE WHEN {} >= '{}' THEN NULL ELSE {} END".format(
+                        ts, first_t, "{}"
+                    )
+                )
+        return result
```

Comment:
Do we have test coverage for all TSA functionalities?

Reply:
Not yet